### PR TITLE
Ecommerce Life Time Value metrics + New Ecommerce section in Visitor Profile

### LINKED ISSUE #2750: LTV analytics - Ecommerce Life Time Value metric in Piwik Visitor Profile and Live APIs
Assuming that an Ecommerce is 100% tied in Piwik, and the business is only internet based, we can hope Piwik to contain all orders and activity of the ecommerce site.

### Lifetime Value for Visitors

* Add LTV metric in the Live APIs 
* Display LTV metric in the Visitor Log 
* Display LTV metric in the Visitor Profile

### Advanced LTV Tracking (to be done later)

With such data over the months, it would be possible to track the Life Time value of a given customer, and come up with reports such as:
- LTV per referrer, websites, campaign and keywords
- LTV per country
- LTV per Custom Variable name and value

To process LTV metric we would need the following
- a way to uniquely identify a customer. For this, we could 
  - use the email address for example. Piwik would then becomes very powerful as a data source. tracker.setUniqueId('client@shop.net');
  - Alternatively, for users who do not wish to use PII for LTV purposes, we would record the hash only, eg. tracker.setHashedUniqueId('client@shop.net') and only the hash would be transmitted over the network.
- on the database side, new log_visit/conversion.unique_id field.
- Archiving
  - LTV archived from log_conversion.revenue
  - The report should always be a running count for current month, and would be summed for process eg. 6 months weighted LTV?
  - processing this report would work even if logs are deleted. However, LTV would have to be processed and aggregated in the running total as often as the logs are deleted in order not to miss any data.
  - Alternatively, a simple first LTV version could simply be the YEARLY report metrics, and then the feature would be that unique customers with NO referrer data lookup the previous referrer data attributed to conversions and copy it in the later conversions, if there is NO referrer more recent. This way we credit the most recent referrer in every conversion, even when first party cookies have been deleted or another computer is used (based off the email adress). This would be a simpler and maybe good enough solution ??
  - It would count the Number of Unique Customers as the sum of Daily orders that were from NEW visitors. If it is from REturning Visitor, Unique Count does not increase, and Purchased products, Revenue from the order, are added to the existing aggregated ecommerce reports.
  - Probably need to add a new metric "New customers" in all ecommerce reports. 
    Example, "March 2011", "Countries report", 
    - 60 German customers bought 110 products for total of 1000E. 40 were NEW visits when they bought. 
    - 100 French customers bought 200 products for total of 1000E. 90 were NEW visits when they bought.
    - DE customers AOV per customer is 1000E/40 = 25E, FR customers is 1000E/90 = 11E.
- If we actually process additional metrics (maybe NOT required see just above), we could process
  - Avg LTV = Sum(ltv) / count(ltv), display in report
  - LT products count, 
  - LT orders count
  - we process a Weighted LTV. 6 months periods avg for example, and ignore all customers who are > 6 months old, in order to show the revenue "speed". Config setting to change this period? Then, we'd need to keep track of  

After this is done, it would be useful to have cost of acquisition of each segment, and process the profit per country, campaign, keyword, etc.

Note: if the company also sells offline and wish to have an accurate LTV report in Piwik, they could still push their offline sales to Piwik using the [Analytics Tracking API](http://piwik.org/docs/tracking-api/). We will publish a doc about this at some point, as I know a few Piwik users are tracking offline already.
- see http://www.jimnovo.com/LTV.htm

### PR DESCRIPTION
**Pull request by mattab**
All Ecommerce metrics shown in the Visitor Profile now apply to "Life time" (before they used to apply to last 100 visits only).

Need to fix system test before merging

Fixes https://github.com/piwik/piwik/issues/2750

### CODE REVIEW

File: plugins/Live/Model.php [hunk 1]
```diff
@@ -509,4 +531,25 @@ private function getWhereClauseAndBind($whereClause, $bindIdSites, $idSite, $per
...
+     */
+    private function getSqlEcommerceConversionsLifeTimeMetricsForIdGoal($ecommerceIdGoal)
+    {
+        $sql = "SELECT
+                    COALESCE(SUM(" . LogAggregator::getSqlRevenue('revenue') . "), 0) as lifeTimeRevenue,
+                    COUNT(*) as lifeTimeConversions,
+                    COALESCE(SUM(" . LogAggregator::getSqlRevenue('items') . "), 0)  as lifeTimeEcommerceItems
+					FROM  " . Common::prefixTable('log_visit') . " AS log_visit
+					    LEFT JOIN " . Common::prefixTable('log_visit') . " AS log_visit_visitors
+					    ON (log_visit.idsite = log_visit_visitors.idsite AND log_visit.idvisitor = log_visit_visitors.idvisitor)
```
Comment (tsteur) [inline plugins/Live/Model.php]: wouldn't we safe one join when matching by idvisitor directly? do we not have the idVisitor in Live API?

File: plugins/Live/Model.php [hunk 2]
```diff
@@ -509,4 +531,25 @@ private function getWhereClauseAndBind($whereClause, $bindIdSites, $idSite, $per
...
     }
+
+    /**
+     * @param $ecommerceIdGoal
+     * @return string
+     */
+    private function getSqlEcommerceConversionsLifeTimeMetricsForIdGoal($ecommerceIdGoal)
+    {
+        $sql = "SELECT
+                    COALESCE(SUM(" . LogAggregator::getSqlRevenue('revenue') . "), 0) as lifeTimeRevenue,
```
Comment (tsteur) [inline plugins/Live/Model.php]: Wondering why we need COALESCE ?

Comment (mattab) [inline plugins/Live/Model.php]: `SUM` returns `NULL` when there are not matching rows (ie. no ecommerce conversions)

Comment (tsteur) [inline plugins/Live/Model.php]: sure, usually just doing it in PHP as usually better readable

File: plugins/Live/Model.php (hunk 1 above)
Comment (mattab) [inline plugins/Live/Model.php]: Good point, will do this

**Pull request merged**